classify-cluster-priority/README.md

```diff
@@ -0,0 +1,85 @@
+# Classifying clusters based on priority
+
+Contents:
+
+1. User stories
+2. Current practice
+3. Proposal
+4. Considerations and explanations
+5. Possible applications
+
+## User stories
+
+- As a cluster admin, I want to easily communicate the purpose and importance of a cluster to those interacting with it, so that they can act accordingly.
+- As Giant Swarm operations staff I want to understand the importance of clusters so I can make appropriate decisions, e. g. when prioritizing problem mitigation in an incident affecting multiple clusters.
+
+## Current practice
+
+- One customer uses the first character of the cluster name to distinguish between production (`p`), development (`d`), and staging (`s`). This customer also has clusters which don't fall into either of these categories and their names start with arbitrary letters.
+- Many customers use specific terms and phrases in the cluster description, e. g. "Production", "test", "testing", "ignore alerts", "prd", "prod", "nonprod".
+- Some customers use labels on the `clusters.cluster.x-k8s.io` resource (the cluster's main resource) indicating the importance, e.g:
+    - `odp/environment: development`
+    - `odp/environment: staging`
+    - `odp/environment: production`.
+
+## Proposal
+
+We introduce a label and a schema for the label values to indicate the relative importance of a cluster.
+
+Label name: `giantswarm.io/service-priority`
+
+The label is to be used on the main resource defining the cluster, usually of resource type `clusters.cluster.x-k8s.io`.
+
+To simplify maintenance and avoid logical conflicts, this label SHOULD NOT be applied on other resources of the cluster. Systems MUST ignore the label used on other resources in the same cluster. This means that it is not possible, for example, to differentiate between the importance of several node pools within the same cluster.
+
+### Well-defined label values
+
+There are three well-defined label values to be used.
+
+- `giantswarm.io/service-priority: critical`: This is the highest priority class. Clusters with this label value are considered the most important ones, relative to other clusters not carrying this label. We expect customers to use this class for clusters serving user-facing applications, production traffic, etc.
+
+- `giantswarm.io/service-priority: important`: Clusters carrying this label are considered less important than `critical`, but still more important than clusters classified as `unimportant`. Typical use cases would be staging clusters, clusters handling batch workloads, or development clusters that cannot be replaced easily by creating a new cluster.
+
+- `giantswarm.io/service-priority: unimportant`: The lowest priority class. Customers are expected to use this class for clusters that isn't relied on, or which can be replaced easily and quickly.
+
+### Absence of the label
+
+The absence of the label SHOULD be interpreted as if the label was present with the value `critical`.
```

Comment:
I would add a kyverno rule that adds a default value. IMO for new clusters, I will default it to `unimportant` as it is usually the case till customer add important services. For existing clusters we can set `critical`

Reply:
Setting a different default value for existing vs. new clusters is something I find too inconsistent and would rather avoid.

In the past we have always dealt with customer clusters in a way that we treated them as production, unless we knew they were not.

I would instead favour making it easy for cluster creators to set the priority, e. g. by having the option it in the cluster creation form in the web UI. The goal here would be to make more users consciously set the priority instead of leaving it at the default.